enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,320 @@
+CRs. Whenever the IPs are updated for a ``DNSName`` CR, the OVN-K master will update the the underlying ``AdressSet``
+referenced by the ACL rule(s) for the corresponding EgressFirewall(s).
+
+OVN-K master will keep track of the TTL (or next lookup time) for each regular DNS name and send a DNS lookup
+query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be performed
+directly on the DNS name as it will not return any IP. Thus, the lookups will be performed on the DNS names
+which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.
+
+### Workflow Description
+
+The workflows for DNS name addition and deletion are explained in this section.
+
+#### Addition
+
+##### Regular DNS name
+* An OpenShift cluster administrator creates an EgressFirewall resource for a Namespace and adds rule(s) containing
+regular DNS name(s).
+* The OVN-K master will create corrresponding `DNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the ``openshift-dns`` Namespace and the Name of the CR will be same as the DNS
+name (barring any trailinng `.`). The ``.spec.isregular`` field of the CR will be set to true, even if it already exists.
+* The OVN-K master will then perform DNS lookup for each of the regular DNS names added to the EgressFirewall rules.
+* The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookup for each of the
+regular DNS names.
+* As these DNS names have corresponding ``DNSName`` CRs, the ``egressfirewall`` plugin will update the ``.status`` of
+the `DNSName` CRs with the DNS name and the corresponding current IPs along with the TTL and the next lookup time based
+on the TTL. However, this update will only take place if there is a change in the exisiting IP
+addresses or next time to lookup or both for the DNS name.
+* The OVN-K master will watch the ``DNSName`` CRs. When the ``.status`` of a ``DNSName`` CR is updated, the OVN-K master
+will update the ``AddressSet`` for the DNS name, which is linked with the ACL rule(s) for the corresponding EgressFirewall(s).
+* The OVN-K master will also receive the response of the DNS lookup query for the DNS name. The OVN-K master will check
+the corresponding ``DNSName`` CR's ``.status`` and if the next lookup time in the status is greater than the next lookup
+time based on the received TTL, then the corresponding CR's ``.status`` will be updated. The corresponding ``AddressSet``
+will also be updated.
+
+##### Wildcard DNS name
+* An OpenShift cluster administrator creates an EgressFirewall resource for a Namespace and adds rule(s) containing
+wildcard DNS name(s).
+* The OVN-K master will create corrresponding `DNSName` CRs for each of the wildcard DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard
```

Comment:
Ok. I will change the namespace to ``openshift-ovn-kubernetes``.

Reply:
Is an EgressFirewall namespaced? What will the parent object of this be? Or will users create these directly?

Reply:
Yes. EgressFirewall is namespaced and it can be created by the cluster admin users. https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html

Reply:
Then I would expect the `EgressFirewallDNSName`, if that's what we are calling it, to be owned by the `EgressFirewall` that caused it to be created, which means it should also be namespaced.